### PR TITLE
feat: quota,shelf rest plugin

### PR DESCRIPTION
**Pull request by Hardikl**
Shelf plugin is fully tested, Quota is in-progress. Logger level will be changed to debug later.

Compared rest metrics with zapi metrics:

Fan:
image

image

Sensor:
image

image

psu: [location is missing]
image

image

image

temperature:
image

image

voltage:
image

image

Quota rest testing is done. 
one difference in rest is, limit value which is not set yet which is unlimited, that field will be omitted in response. 

qtree:
image

image

Quota:
disk_limit: need to convert byte to kilobyte in rest
image

disk_used:
image

disk_used_pct_disk_limit:
image

disk_used_pct_soft_disk_limit:
image

soft_disk_limit:
image

file_limit:
image

files_used:
image

files_used_pct_file_limit:
image

files_used_pct_soft_file_limit:
image

soft_file_limit:
image

### CODE REVIEW

File: cmd/collectors/rest/plugins/qtree/qtree.go [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+
+	for _, quota := range results[1].Array() {
+		var tree string
+
+		if !quota.IsObject() {
+			my.Logger.Error().Str("type", quota.Type.String()).Msg("Quota is not an object, skipping")
+			return nil, errors.New(errors.ERR_NO_INSTANCE, "quota is not an object")
+		}
+
+		if quota.Get("qtree").Exists() {
```
Comment (rahulguptajss) [inline cmd/collectors/rest/plugins/qtree/qtree.go:159]: may be qtree.name check is better

File: cmd/collectors/rest/plugins/qtree/qtree.go [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+			tree = quota.Get("qtree.name").String()
+		}
+		volume := quota.Get("volume.name").String()
+		vserver := quota.Get("svm.name").String()
+		quotaIndex := quota.Get("index").String()
+
+		// If quota-type is not a tree, then skip
+		if quota.Get("type").String() != "tree" {
+			continue
```
Comment (rahulguptajss) [inline cmd/collectors/rest/plugins/qtree/qtree.go:168]: add a trace log

File: cmd/collectors/rest/plugins/shelf/shelf.go [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+		instanceKeys := exportOptions.NewChildS("instance_keys", "")
+		instanceKeys.NewChildS("", "shelf")
+		instanceKeys.NewChildS("", "channel")
+
+		// artificial metric for status of child object of shelf
+		my.data[attribute].NewMetricUint8("status")
+
+		for _, c := range childObj {
+
+			metricName, display := collector.ParseMetricName(c)
```
Comment (rahulguptajss) [inline cmd/collectors/rest/plugins/shelf/shelf.go:123]: you can move templating method to util and use that here.

File: cmd/collectors/rest/plugins/qtree/qtree.go [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+		my.Logger.Error().Stack().Err(err).Msg("connecting")
+		return err
+	}
+
+	if err = my.client.Init(5); err != nil {
+		return err
+	}
+
+	my.query = "api/storage/quota/reports"
+	my.Logger.Info().Msg("plugin connected!")
```
[the remark this answers is not in the document]
Comment (Hardikl) [reply cmd/collectors/rest/plugins/qtree/qtree.go:68]: removed

File: cmd/collectors/rest/plugins/qtree/qtree.go (hunk 1 above)
Comment (Hardikl) [reply cmd/collectors/rest/plugins/qtree/qtree.go:159]: done.

File: cmd/collectors/rest/plugins/qtree/qtree.go (hunk 2 above)
Comment (Hardikl) [reply cmd/collectors/rest/plugins/qtree/qtree.go:168]: done

File: cmd/collectors/rest/plugins/shelf/shelf.go (hunk 3 above)
Comment (Hardikl) [reply cmd/collectors/rest/plugins/shelf/shelf.go:123]: done, moved to util package.

File: pkg/util/util.go [hunk 5]
```diff
@@ -327,3 +327,34 @@ func Intersection(a []string, b []string) ([]string, []string) {
 
 	return matches, misses
 }
+
+// ParseMetric parses display name and type of field from the raw name of the metric as defined in (sub)template.
+// Users can rename a metric with "=>" (e.g. some_long_metric_name => short).
+// Trailing "^" characters are ignored/cleaned as they have special meaning in some collectors.
+func ParseMetric(rawName string) (string, string, string) {
```
Comment (rahulguptajss) [inline pkg/util/util.go:334]: could u add hyphon to underscore conversion in this method.

Comment (Hardikl) [reply pkg/util/util.go:334]: added.

**Review by rahulguptajss — APPROVED** [review]

**Pull request merged**